db/crud.go

```diff
@@ -882,278 +1072,109 @@ func (db *Database) IsIllegalConflict(doc *Document, parentRevID string, deleted
 	return true
 }
 
-// Function type for the callback passed into updateAndReturnDoc
-type updateAndReturnDocCallback func(*Document) (resultBody Body, resultAttachmentData AttachmentData, updatedExpiry *uint32, resultErr error)
+func (db *Database) documentUpdateFunc(docExists bool, doc *Document, allowImport bool, previousDocSequence uint64, unusedSequences []uint64, callback updateAndReturnDocCallback, expiry uint32) (retDoc *Document, retSyncFuncExpiry *uint32, retBody Body, retNewRevID string, retStoredBody Body, retOldBodyJSON string, retDocSequence uint64, retUnusedSequences []uint64, changedPrincipals []string, changedRoleUsers []string, err error) {
```

Comment:
I think renaming 'changedPrincipals' and 'changedRoleUsers' to 'changedAccessPrincipals' and 'changedRoleAccessUsers' makes it more transparent what these represent (i.e. the relationship to the sync function is clearer).  It's a bit more wordy, but I think it makes the code quite a bit more self-describing.